datashader/glyphs/points.py

```diff
@@ -159,9 +159,9 @@ def _perform_extend_points(i, sx, tx, sy, ty, xmin, xmax, ymin, ymax, xs, ys, *a
             if (xmin <= x <= xmax) and (ymin <= y <= ymax):
                 xx = int(x_mapper(x) * sx + tx)
                 yy = int(y_mapper(y) * sy + ty)
-                xi, yi = (xx - 1 if x == xmax else xx,
-                          yy - 1 if y == ymax else yy)
 
+                xi, yi = (xx - 1 if x >= (xmax-1e-15) else xx,
+                          yy - 1 if y >= (ymax-1e-15) else yy)
```

Comment:
Fundamentally, we aren't meant to get to this location unless x<=xmax and y<=ymax, so can we be more direct about this?  E.g. can we calculate the integer xx and yy values that are out of bounds, and then test against xx and yy rather than x and y with an epsilon?  That seems like a more direct way to enforce that the results are in bounds, with no arbitrary constants.

Reply:
e.g. 

```
xi, yi = (xxmax - 1 if xx >= xxmax else xx,
          yymax - 1 if yy >= yymax else yy)
```

Seems like this will work as long as we can calculate xxmax and yymax.

Reply:
e.g. 

```
xi, yi = (xxmax - 1 if xx >= xxmax else xx,
          yymax - 1 if yy >= yymax else yy)
```

Seems like this will work as long as we can calculate xxmax and yymax.

Reply:
Definitely slower which is why I didn't suggest it. We could compute that once and pass it in as arguments, just seems awkward all around. Alternatively we could check the shape of the first aggregate array in aggs_and_cols, but that would probably break with the argument expansion ast manipulation code.

Reply:
It does seem awkward, but an epsilon doesn't seem like a viable alternative unless it were a quantity expressed in terms of the fundamental floating-point precision, and not just an arbitrary number that will affect data ranges of different magnitudes differently.

Reply:
1e-15 is one order of magnitude larger than sys.float_info.epsilon. That said I'm not actually arguing this should be merged. I think we will just have to pass in the array bounds.

Reply:
Passing in the array bounds is ugly but I don't see any viable alternative either.

Reply:
Done now.